Oxidize commutative cancellation

sbrandhsn opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [x] I have added the tests to cover my changes.
- [x] I have updated the documentation accordingly.
- [x] I have read the CONTRIBUTING document.
-->

### Summary

fixes https://github.com/Qiskit/qiskit/issues/12270

### Details and comments

This replaces the Python implementation of `CommutativeCancellation` with a Rust implementation:

- As the Rust implementation does not use the `PassManager` to run its prerequisite `CommutationAnalysis`, the test case has been `test_callback_with_pass_requires` fixed. We might be able to remove that test case as the check is already covered in https://github.com/Qiskit/qiskit/blob/52733ae09f7ac9d2d5e2a674970ff95301d34bd6/test/python/transpiler/test_pass_scheduler.py#L111
- The global phase was probably incorrectly set in the Python implementation in https://github.com/Qiskit/qiskit/blob/f4ca088644b4461b856b4b49c33e23b04a995019/qiskit/transpiler/passes/optimization/commutative_cancellation.py#L195, I changed this in the Rust implementation
- Currently, control-flow operations are handled in Python - I'm looking into a Rust implementation for that too

### Speedups (this branch vs Qiskit 1.2)

![image](
![clifford_circuits](https://github.com/user-attachments/assets/d66401df-49ec-4b4f-ba02-80406cab4777)
![random_circuits](https://github.com/user-attachments/assets/868d2344-ad94-4749-9ae6-c794c833f36f)
![rotations_circuits](https://github.com/user-attachments/assets/ef769fbd-7019-4b0b-a4a1-cdb153adf3bd))

Top to bottom Clifford circuits, random circuits, random circuits with rotation gates only

Up to 5.5x speedup for Clifford circuits where the commutation relation of two gates is always determined by a look-up in an a priori constructed commutation library. Up to 7.0 speedup for random circuits as generated by `random_circuit` and up to 8x speedup when restricting that random circuit to rotational gates only. Note this benchmark compares end-to-end performance as given by
```
            pm = PassManager([CommutativeCancellation(target=be.target)])
            start_ts = time.time()
            pm.run(qc)
            end_ts = time.time()
```
instead of only comparing a python implementation of `CommutativeCancellation` with underlying Rust code vs a rust implementation of `CommutativeCancellation` with the same underlying code.

mtreinish commented on crates/accelerate/src/commutation_cancellation.rs:
```diff
@@ -0,0 +1,285 @@
+) -> PyResult<()> {
+    let basis: HashSet<String> = if let Some(tar) = target {
+        HashSet::from_iter(tar.operation_names().map(String::from))
+    } else if let Some(basis) = basis_gates {
```
While we can do this from Rust, this actually adds overhead to every call of the pass, while if we do this in Python we only have the cost once per initialization. Since we run the cancellation pass as part of the optimization loop this could be a performance hit. I would just do the set collection for the target in Python and only pass `basis_gates` to the rust function.

mtreinish commented on crates/accelerate/src/commutation_cancellation.rs:
```diff
@@ -0,0 +1,285 @@
+    };
+
+    let _var_z_map: HashMap<&str, StandardGate> =
+        HashMap::from([("rz", RZGate), ("p", PhaseGate), ("u1", U1Gate)]);
```
I wonder if it would make more sense to have this as a `static VAR_Z_MAP: [(&str, StandardGate); 3]`. We only are doing 3 string comparisons of at most 2 characters it feels like a hashmap here is more expensive that just checking the 3 strings.

mtreinish commented on crates/accelerate/src/commutation_cancellation.rs:
```diff
@@ -0,0 +1,285 @@
+    let mut single_q_cancellation_sets: HashMap<(String, Qubit, usize), Vec<NodeIndex>> =
+        HashMap::new();
+    let mut two_q_cancellation_sets: HashMap<(String, Qubit, Qubit, usize, usize), Vec<NodeIndex>> =
+        HashMap::new();
```
We're working solely with `StandardGates` here instead of storing a `String` can't we store the `StandardGate`? We'd have to change `_gates` too but I think that's ok. We also can make this all static more easily by using `StandardGate` because it's a very fast u8 comparison under the covers.

mtreinish replied:
I think we can define better types for these that are easier to work, a bit faster, and less error prone. Something like:

```rust
    #[derive(Hash, Eq, PartialEq, Debug)]
    enum GateOrRotation {
        Gate(StandardGate),
        ZRotation,
        XRotation,
    }           
    #[derive(Hash, Eq, PartialEq, Debug)]
    struct CancellationSetKey {
        gate: GateOrRotation,
        qubits: SmallVec<[Qubit; 2]>,
        com_set_index: usize,
        second_index: Option<usize>,
    }
 
    let mut cancellation_sets: HashMap<CancellationSetKey, Vec<NodeIndex>> = HashMap::new();
```

Would let us use one hashmap instead of two and maybe simplify the logic of generating the cancellation sets. Although we could easily do two hashmaps here too if that's preferable and just have two key types. The struct will be easier to work with because everything is named access and the enum will let us use `StandardGate` instead of strings.

mtreinish commented on crates/accelerate/src/commutation_cancellation.rs:
```diff
@@ -0,0 +1,285 @@
+                    }
+                })
+        }
+    });
```
I was curious and sketched out using a parallel iterator chain here to build the cancellation sets hashmap using the above suggestion on changing the types:

```rust
    let cancellation_sets: HashMap<CancellationSetKey, Vec<NodeIndex>> = (0..dag.num_qubits())
        .into_par_iter()
        .filter_map(|qubit| {
            let wire = Qubit(qubit as u32);
            if let Some(wire_commutation_set) = commutation_set.get(&Wire::Qubit(wire)) {
                Some(wire_commutation_set
                    .iter()
                    .enumerate()
                    .filter_map(|(com_set_index, com_set)| {
                        // This ensures that we only have DAGOPNodes in the current com_set, yuck...
                        if let NodeType::Operation(_node0) = &dag.dag[*com_set.first().unwrap()] {
                            Some(com_set.iter().filter_map(|node| {
                                let op = match &dag.dag[*node] {
                                    NodeType::Operation(instr) => instr,
                                    _ => panic!("Unexpected type in commutation set."),
                                };
                                let num_qargs = dag.get_qargs(op.qubits).len();
                                // no support for cancellation of parameterized gates
                                if op
                                    .params_view()
                                    .iter()
                                    .all(|p| !matches!(p, Param::ParameterExpression(_)))
                                {
                                    let op_name = op.op.name().to_string();
                                    let gate = op.op.standard_gate();
                                    if num_qargs == 1usize && GATES.contains(&gate) {
                                        let key = CancellationSetKey {
                                            gate: GateOrRotation::Gate(gate),
                                            qubits: smallvec![wire],
                                            com_set_index,
                                            second_index: None,
                                        };
                                        Some((key, *node))
                                    } else if num_qargs == 1usize && _z_rotations.contains(op_name.as_str()) {
                                        let key = CancellationSetKey {
                                            gate: GateOrRotation::Z_ROTATION,
                                            qubits: smallvec![wire],
                                            com_set_index,
                                            second_index: None,
                                        };
                                        Some((key, *node))
                                    } else if num_qargs == 1usize && _x_rotations.contains(op_name.as_str()) {
                                        let key = CancellationSetKey {
                                            gate: GateOrRotation::X_ROTATION,
                                            qubits: smallvec![wire],
                                            com_set_index,
                                            second_index: None,
                                        };
                                        Some((key, *node))
                                    }
                                    // Don't deal with Y rotation, because Y rotation doesn't commute with
                                    // CNOT, so it should be dealt with by optimized1qgate pass
                                    else if num_qargs == 2usize
                                        && dag.get_qargs(op.qubits).first().unwrap() == &wire
                                    {
                                        let second_qarg = dag.get_qargs(op.qubits)[1];
                                        let key = CancellationSetKey {
                                            gate: GateOrRotation::Gate(gate),
                                            qubits: smallvec![wire, second_qarg],
                                            com_set_index,
                                            second_index: Some(*node_indices.get(&(*node, Wire::Qubit(second_qarg))).unwrap()),
                                        };
                                        Some((key, *node))
                                    } else {
                                        None
                                    }
                                } else {
                                    None
                                }
                        }).collect::<Vec<(CancellationSetKey, NodeIndex)>>())
                    } else {
                        None
                    }
                })
                .flat_map(|entries| entries.into_iter()))
            } else {
                None
            }
        })
        .fold(|| HashMap::new(), |map, item| {
            for (key, value) in item {
                map.entry(key).and_modify(|indices: &mut Vec<NodeIndex>| indices.push(value)).or_insert_with(|| vec![value]);
            }
            map
        })
        .reduce(|| HashMap::new(), |outer_map, inner_map| {
            for (key, value) in inner_map.iter_mut() {
                outer_map.entry(*key).and_modify(|indices| indices.append(value)).or_insert_with(|| *value);
            }
            outer_map
        });
```

This doesn't fully compile yet because I haven't changed the logic below it to work with a single hashmap yet. But it at least gets passed the type checking (which probably means the borrow checker will hate it :D). I'm skeptical doing this in parallel actually helps performance here though because it results in several intermediate allocations that could be avoided. It was more just an exercise for me to see how it could be done. But this might be useful as a reference for other changes so I decided to post it even though it's not complete.

mtreinish commented on crates/accelerate/src/commutation_cancellation.rs:
```diff
@@ -0,0 +1,285 @@
+                    || node_qargs.len() > 1
+                    || &node_qargs[0] != run_qarg
+                {
+                    panic!("internal error");
```
While not helpful for debugging at all this was a Python space error before, I think it's probably worth returning a `RuntimeError` to python here.

mtreinish commented on crates/accelerate/src/commutation_cancellation.rs:
```diff
@@ -0,0 +1,285 @@
+                        Some(Param::Float(f)) => *f,
+                        _ => panic!(
+                            "Rotational gate with parameter expression encoutned in cancellation"
+                        ),
```
This should be returned to Python as a runtime error so it's "user actionable"

mtreinish commented on crates/accelerate/src/commutation_cancellation.rs:
```diff
@@ -0,0 +1,285 @@
+                total_angle += node_angle;
+
+                if let Some(definition) = node_op.op.definition(node_op.params_view()) {
+                    //TODO check for PyNone global phase?
```
Global phase can't be `None` in rust. It's always a Param type: https://github.com/Qiskit/qiskit/blob/main/crates/circuit/src/circuit_data.rs#L104

mtreinish commented on crates/accelerate/src/commutation_cancellation.rs:
```diff
@@ -0,0 +1,285 @@
+            } else if cancel_key.0 == X_ROTATION {
+                &RXGate
+            } else {
+                panic!("impossible case!");
```
If this really impossible then I'd use `unreachable!()` here.  But this is also a really good case to have the `enum`  I described above. because it makes all the possible values here very explicit.

sbrandhsn replied:
I changed this to a Python error. It was in the original code and made me chuckle - hence I thought it is worth keeping it like this. :-) Let me know if you want it to be `unreachable!` instead.

mtreinish commented on crates/accelerate/src/commutation_cancellation.rs:
```diff
@@ -0,0 +1,285 @@
+
+/// Wrap angle into interval [-π,π). If within atol of the endpoint, clamp to -π
+#[inline]
+fn mod_2pi(angle: f64, atol: f64) -> f64 {
```
I feel like we have a function like this in a bunch of places already. One not just call it instead of copying it?

mtreinish commented on qiskit/transpiler/passes/optimization/commutative_cancellation.py:
```diff
@@ -70,12 +64,11 @@ def __init__(self, basis_gates=None, target=None):
         # build a commutation checker restricted to the gates we cancel -- the others we
         # do not have to investigate, which allows to save time
-        commutation_checker = CommutationChecker(
+        self.commutation_checker = CommutationChecker(
```
Just so people don't assume this is publicly accessible:

```suggestion
        self._commutation_checker = CommutationChecker(
```

mtreinish commented on crates/accelerate/src/commutation_cancellation.rs:
```diff
@@ -0,0 +1,285 @@
+            /*
+                if np.mod(total_angle, (2 * np.pi)) < _CUTOFF_PRECISION:
+                    dag.remove_op_node(run[0])
+            */
```
In it's current form this isn't needed since you're removing all the gates regardless of the angle so I would just delete the comment. If this is needed we'd have to update the removal loop to account for this.

```suggestion
```

sbrandhsn replied:
Yeah, I did not understand why it would be needed in the original code and thought it may be related to numerical instability. I will remove it.

mtreinish commented on crates/accelerate/src/commutation_cancellation.rs:
```diff
@@ -0,0 +1,285 @@
+
+            let run_qarg = dag.get_qargs(run_op.qubits).first().unwrap();
+            let mut total_angle: f64 = 0.0f64;
+            let mut total_phase: f64 = 0.0f64;
```
You don't need the suffix because you've set a type annotation here.
```suggestion
            let mut total_angle: f64 = 0.0;
            let mut total_phase: f64 = 0.0;
```

mtreinish commented on crates/accelerate/src/commutation_analysis.rs:
```diff
@@ -50,7 +50,7 @@ const MAX_NUM_QUBITS: u32 = 3;
 ///     node_indices = {(0, 0): 0, (1, 0): 3, (2, 0): 1, (3, 0): 1, (4, 0): 2}
 ///
-fn analyze_commutations_inner(
+pub fn analyze_commutations_inner(
```
I'm fine with making this `pub` but `pub(crate)` is also an option since this is only going to be used inside the same crate for now (this will likely change in the medium term towards Qiskit 2.0).

mtreinish reviewed (commented):
Thanks for doing this, this has the potential to improve performance quite a bit, not only by speeding up the cancellation, but also by avoiding all the python overhead for the analysis. I've done a quick first pass over the code. I left a few inline comments, besides those the big thing I think we want to be careful for here is making sure we don't panic needlessly and instead return Python exceptions when there is a possible error condition. Right now the code seems a bit too likely to just call `panic!()` which won't be use actionable or catchable and also exits hard.

mtreinish commented:
Oh also can you please rebase this on main so the commit history is just the local changes. While the final diff is ok, it makes the commit message and authors list for this PR the sum of all the commits on the branch, which is much larger than the actual work being done here.

sbrandhsn committed:
fix

sbrandhsn committed:
fmt

sbrandhsn committed:
comments from code review

mtreinish commented on crates/accelerate/src/commutation_cancellation.rs:
```diff
@@ -0,0 +1,291 @@
+
+                    if let Some(definition) = node_op.op.definition(node_op.params_view()) {
+                        total_phase += match definition.global_phase() {Param::Float(f) => f, _ => panic!("PackedInstruction with definition has no global phase set as floating point number")};
+                    }
```
As the logic is right now this should return a python error:

```suggestion
                        total_phase += match definition.global_phase() {
                            Param::Float(f) => f,
                             _ => return Err(QiskitError::new_err(format!("PackedInstruction with definition has no global phase set as floating point number")))
                        };
                    }
```

Although realistically if you make my above suggestion about how to get the definition directly from `StandardGate` I think you can change this to just use `unreachable!()` since the generator code for definitions of standard gates never has a parameterized phase, it's always a float `0.0`. Something like:

```suggestion
                        let Param::Float(new_phase) = definition.global_phase() else { unreachable!() };
                        total_phase += new_phase
```
(although maybe set an error message saying standard gate phase is never parameterized)

would do the trick. But we can't make this change unless we're working with a `StandardGate` when we call `definition()`.

jakelishman replied:
Unless I'm missing something: one of Rz or P (whichever isn't exactly equal to u1) can have a parametric phase if the angle is parametric. `GlobalPhaseGate` definitely can (if the angle is parametric).

mtreinish replied:
Ah good point, I thought there was one, but I scanned through the match statement yesterday and missed it on rzgate. But regardless this statement still holds since the angle is always a float at this point

mtreinish commented on crates/accelerate/src/commutation_cancellation.rs:
```diff
@@ -0,0 +1,291 @@
+                    };
+
+                    if let Some(definition) = new_node.op.definition(new_node.params_view()) {
+                        match definition.global_phase() {Param::Float(f) => *f, _ => panic!("PackedInstruction with definition has no global phase set as floating point number")}
```
Same comment above with the global phase handling.

mtreinish commented on crates/accelerate/src/commutation_cancellation.rs:
```diff
@@ -0,0 +1,291 @@
+                        match node_op.params_view().first() {
+                            Some(Param::Float(f)) => *f,
+                            _ => return Err(QiskitError::new_err(format!(
+                                "Rotational gate with parameter expression encoutned in cancellation {:?}",
```
```suggestion
                                "Rotational gate with parameter expression encountered in cancellation {:?}",
```

mtreinish commented on crates/accelerate/src/commutation_cancellation.rs:
```diff
@@ -0,0 +1,291 @@
+                .for_each(|(com_set_idx, com_set)| {
+                    // This ensures that we only have DAGOPNodes in the current com_set, yuck...
+                    if let NodeType::Operation(_node0) = &dag.dag[*com_set.first().unwrap()] {
+                        com_set.iter().for_each(|node| {
```
Besides removing the commentary on the syntax, I'm not sure exactly what the intent is with this code. Are you trying to filter sets here that don't start with an `Operation` node? There are a couple of options if so which might  be more succinct, such as doing:

```rust
for (com_set_idx, com_set) in  wire_commutation_set.iter().enumerate() {
    if Some(node) = com_set.first() {
        if !matches!(NodeType::Operation(_), dag.dag[node]) {
            continue
        }
    } else {
        continue
    }
```

or if you'd like to stick with an iterator you can do:

```rust
.filter(|(com_set_idx, com_set)| {
    if Some(node) = com_set.first() {
        if matches!(NodeType::Operation(_), dag.dag[node]) {
            true
        } else {
           false
        }
    } else {
        false
    }
})
.for_each(|(com_set_idx, com_set)| {
```

Although this depends on what the logic you're trying to do is. If you don't want to tolerate a single non-op node in the commutation set then there is different logic I'd probably use. But either way I'd suggest making either change here because besides not being a potential source of a panic it also removes one layer of nesting from the for loop here.

sbrandhsn replied:
Whoops, forgot to remove the comment - sorry! `com_set` only contains commuting entries if the first element is an operation. I don't think this logic is ideal but I also didn't want to refactor it to something that may take more time. I changed the code to the first example above.

mtreinish commented on crates/accelerate/src/commutation_cancellation.rs:
```diff
@@ -0,0 +1,291 @@
+                            let op = match &dag.dag[*node] {
+                                NodeType::Operation(instr) => instr,
+                                _ => panic!("Unexpected type in commutation set."),
+                            };
```
Related to above would it just make more sense to have this be a skip condition instead of a panic and remove the `first()` logic. So if we encounter a `non-Operation` node type we just ignore this set?

sbrandhsn replied:
We would deviate from the original code by doing so, but this is fine for me. Essentially, these `panic!` instructions check whether the source of `com_set` follows the contract/structure assumed in `CommutativeCancellation`. I think I would like to make the format checks explicit instead of simply passing over malformed `com_set` - what do you think?

mtreinish commented on crates/accelerate/src/commutation_cancellation.rs:
```diff
@@ -0,0 +1,291 @@
+                                .params_view()
+                                .iter()
+                                .all(|p| !matches!(p, Param::ParameterExpression(_)))
+                            {
```
There is `op.is_parameterized()` here which you can probably use here instead with:

```suggestion
                            if !op.is_parameterized() {
```

mtreinish commented on crates/accelerate/src/commutation_cancellation.rs:
```diff
@@ -0,0 +1,291 @@
+                    // This ensures that we only have DAGOPNodes in the current com_set, yuck...
+                    if let NodeType::Operation(_node0) = &dag.dag[*com_set.first().unwrap()] {
+                        com_set.iter().for_each(|node| {
+                            let op = match &dag.dag[*node] {
```
Personally I'd name this `instr` instead of `op` because it's a bit confusing to have `op` as type `PackedInstruction`. When I saw `op` below I thought for a second it was a `PackedOperation` or something with the `Operation` trait implemented.

mtreinish commented on crates/accelerate/src/commutation_cancellation.rs:
```diff
@@ -0,0 +1,291 @@
+                                .all(|p| !matches!(p, Param::ParameterExpression(_)))
+                            {
+                                if let Some(op_gate) = op.op.try_standard_gate() {
+                                    if num_qargs == 1usize && SUPPORTED_GATES.contains(&op_gate) {
```
You don't actually need the `usize` suffix here, the compiler is smart enough to infer it because `num_qargs` is a usize.

```suggestion
                                    if num_qargs == 1 && SUPPORTED_GATES.contains(&op_gate) {
```

mtreinish commented on crates/accelerate/src/commutation_cancellation.rs:
```diff
@@ -0,0 +1,291 @@
+                    _ => panic!("Unexpected type in commutation set run."),
+                };
+
+                let run_qarg = dag.get_qargs(run_op.qubits).first().unwrap();
```
It's a slice so you can just get this with an index:
```suggestion
                let run_qarg = dag.get_qargs(run_op.qubits)[0];
```

mtreinish replied:
I'd be inclined to just simplify this as:

```suggestion
                let run_qarg = match &dag.dag[*cancel_set.first().unwrap()] {
                    NodeType::Operation(instr) => dag.get_qargs(instr.qubits)[0],
                    _ => panic!("Unexpected type in commutation set run."),
                };
```

mtreinish commented on crates/accelerate/src/commutation_cancellation.rs:
```diff
@@ -0,0 +1,291 @@
+                    } else if EIGHTH_TURNS.contains(&node_op_name) {
+                        PI / 4.0
+                    } else {
+                        panic!("Angle for operation {node_op_name} is not defined")
```
I feel like this should be a python error because it feels like it's user actionable.

mtreinish commented on crates/accelerate/src/commutation_cancellation.rs:
```diff
@@ -0,0 +1,291 @@
+                    let new_node = match &dag.dag[new_index] {
+                        NodeType::Operation(instr) => instr,
+                        _ => panic!("Unexpected type in commutation set run."),
+                    };
```
Why do you need to do this if we defined `new_op` above? Can't we just get the definition and global phase directly from `new_op` with `new_op.definition(&[total_angle])`?

mtreinish commented on test/python/transpiler/test_commutative_cancellation.py:
```diff
@@ -76,7 +77,7 @@ def test_all_gates(self):
         expected.append(RZGate(2.0), [qr[0]])
         expected.rx(1.0, qr[0])
-
+        expected.global_phase = 0.5
```
I might have missed this in the PR summary message, but was this caused by the change in logic you made around the global phase handling?

sbrandhsn replied:
Yes. The previous code would only consider the last replaced gate when setting the new `global_phase`. I think this was essentially a typo of 
```
global_phase = new_phase
```
instead of 
```
global_phase += new_phase
```

mtreinish reviewed (commented):
Thanks for the quick updates. I took another pass through the code after the last updates. This is looking a lot better now and much less error prone using an enum and rust types instead of string comparisons for everything. I'm still a bit concerned how many `panic!()` calls there are though, I feel like a fair number of them can be avoided though, and either eliminated with some logical changes or replaced with Python errors. I left some inline comments with suggestions on how to do this.

sbrandhsn committed:
comments from code review

sbrandhsn committed:
Merge branch 'main' into oxidize-commutative-cancellation

sbrandhsn committed:
Update lib.rs

mtreinish commented on crates/accelerate/src/commutation_cancellation.rs:
```diff
@@ -0,0 +1,280 @@
+use qiskit_circuit::Qubit;
+use rustworkx_core::petgraph::stable_graph::NodeIndex;
+use smallvec::{smallvec, SmallVec};
+use std::f64::consts::PI;
```
Small nit, just grouping the imports here between stdlib, external crates, and internal crates:

```suggestion
use std::f64::consts::PI;

use hashbrown::{HashMap, HashSet};
use pyo3::prelude::*;
use pyo3::{pyfunction, pymodule, wrap_pyfunction, Bound, PyResult, Python};
use rustworkx_core::petgraph::stable_graph::NodeIndex;
use smallvec::{smallvec, SmallVec};

use qiskit_circuit::dag_circuit::{DAGCircuit, NodeType, Wire};
use qiskit_circuit::operations::StandardGate::{
    CXGate, CYGate, CZGate, HGate, PhaseGate, RXGate, RZGate, SGate, TGate, U1Gate, XGate, YGate,
    ZGate,
};
use qiskit_circuit::operations::{Operation, Param, StandardGate};
use qiskit_circuit::Qubit;

use crate::commutation_analysis::analyze_commutations_inner;
use crate::commutation_checker::CommutationChecker;
use crate::{euler_one_qubit_decomposer, QiskitError};
```

mtreinish commented on crates/accelerate/src/commutation_cancellation.rs:
```diff
@@ -0,0 +1,280 @@
+                } else {
+                    continue;
+                }
+                com_set.iter().for_each(|node| {
```
Personally I'd probably just use a normal for loop for this instead of `for_each()` but it functionally doesn't really matter. I typically reserve `for_each()` when I have a longer iterator chain that has maps and stuff and then need to iterate over that at the end.

If you did use a normal for loop you could reduce the nesting one layer and do:

```rust
if instr.is_parameterized() {
    continue;
}
```

which also would match the python logic more closely.

mtreinish commented on crates/accelerate/src/commutation_cancellation.rs:
```diff
@@ -0,0 +1,280 @@
+                            // Don't deal with Y rotation, because Y rotation doesn't commute with
+                            // CNOT, so it should be dealt with by optimized1qgate pass
+                            if num_qargs == 2
+                                && dag.get_qargs(instr.qubits).first().unwrap() == &wire
```
```suggestion
                                && dag.get_qargs(instr.qubits)[0] == &wire
```

mtreinish commented on crates/accelerate/src/commutation_cancellation.rs:
```diff
@@ -0,0 +1,280 @@
+                    }
+                }
+            }
+            if matches!(cancel_key.gate, GateOrRotation::ZRotation) && z_var_gate.is_none() {
+                continue;
+            }
```
In the python version this was independent of `cancel_set.len() > 1`, does this matter? I guess probably not because if we had 1 gate in the cancel set we'd skip regardless of what the key was.

sbrandhsn replied:
Yes, I agree

mtreinish commented on crates/accelerate/src/commutation_cancellation.rs:
```diff
@@ -0,0 +1,280 @@
+            if matches!(cancel_key.gate, GateOrRotation::ZRotation)
+                || matches!(cancel_key.gate, GateOrRotation::XRotation)
+            {
+                let run_qarg = match &dag.dag[*cancel_set.first().unwrap()] {
```
```suggestion
                let run_qarg = match &dag.dag[*cancel_set[0]] {
```

mtreinish commented on crates/accelerate/src/commutation_cancellation.rs:
```diff
@@ -0,0 +1,280 @@
+                continue;
+            }
+            if matches!(cancel_key.gate, GateOrRotation::ZRotation)
+                || matches!(cancel_key.gate, GateOrRotation::XRotation)
```
I think this syntax works to make it a little more concise: 
```suggestion
            if matches!(cancel_key.gate, GateOrRotation::ZRotation | GateOrRotation::XRotation)
```

mtreinish commented on crates/accelerate/src/commutation_cancellation.rs:
```diff
@@ -0,0 +1,280 @@
+                    if node_op
+                        .extra_attrs
+                        .as_deref()
+                        .is_some_and(|attr| attr.condition.is_some())
```
```suggestion
                    if node_op.condition().is_some()
```

mtreinish commented on crates/accelerate/src/commutation_cancellation.rs:
```diff
@@ -0,0 +1,280 @@
+                        || node_qargs.len() > 1
+                        || node_qargs[0] != run_qarg
+                    {
+                        return Err(QiskitError::new_err("internal error"));
```
For "backwards compatibility" (whatever that means in this case) we should return a `PyRuntimeError` here since that's what the python version returned. But realistically I find it super unlikely anyone was depending on this exact exception being raised. I think realistically we might be able to remove this check and error completely since none of these conditions should be possible.

mtreinish commented on crates/accelerate/src/commutation_cancellation.rs:
```diff
@@ -0,0 +1,280 @@
+                    } else if EIGHTH_TURNS.contains(&node_op_name) {
+                        Ok(PI / 4.0)
+                    } else {
+                        Err(QiskitError::new_err(format!(
```
Similarly for backwards compatibility this would be a `PyRuntimeError` not a `QiskitError` but I also don't think this is something we could trigger. Ideally we maybe should look at using a more rust native structure that doesn't have a fall through possibility like this. But I don't have any concrete ideas so we can just leave this error path as is.

mtreinish commented on crates/accelerate/src/commutation_cancellation.rs:
```diff
@@ -0,0 +1,280 @@
+                    for &c_node in &cancel_set[0..(cancel_set.len() / 2) * 2] {
+                        dag.remove_op_node(c_node);
+                    }
+                }
```
I think this needs a continue here because this is an xor with the other two gate fields. In the python this was if gate else if rotation.
```suggestion
                }
                continue;
```

mtreinish commented on crates/accelerate/src/commutation_cancellation.rs:
```diff
@@ -0,0 +1,280 @@
+                    &RXGate
+                } else {
+                    return Err(QiskitError::new_err("impossible case!"));
+                };
```
I think using a match here would be more natural:
```suggestion
                let new_op = match cancel_key.gate
                    GateOrRotation::ZRotation => z_var_gate.unwrap(),
                    GateOrRotation::XRotation) => &RXGate,
                    _ => unreachable!()
                };
```

I also switched this to an `unreachable!()` because while reviewing the current form of the PR it is obvious that this branch is never reachable because  after re-reading the code I realize the else arm here couldn't ever be reached since this all in a block that gate matches either rotation variant of the enum.

mtreinish reviewed (commented):
I think this almost ready now, the current state of this PR is looking much better. Thanks for making all the updates. I have another round of inline comments, but most are pretty small nits and mechanical things to make the code a bit more concise.

sbrandhsn committed:
Apply suggestions from code review

Co-authored-by: Matthew Treinish <mtreinish@kortar.org>

sbrandhsn reviewed (commented):
Thanks for the feedback! :-)

sbrandhsn committed:
code review

mtreinish commented on crates/circuit/src/dag_circuit.rs:
```diff
@@ -6228,7 +6228,7 @@ impl DAGCircuit {
         new_gate: (StandardGate, &[f64]),
         old_index: NodeIndex,
-    ) {
+    ) -> NodeIndex {
```
```suggestion
    ) {
```

mtreinish commented on crates/circuit/src/dag_circuit.rs:
```diff
@@ -6255,6 +6255,7 @@ impl DAGCircuit {
         self.dag.add_edge(parent_index, new_index, weight.clone());
         self.dag.add_edge(new_index, old_index, weight);
         self.dag.remove_edge(edge_index);
+        new_index
```
```suggestion
```

mtreinish commented on crates/circuit/src/dag_circuit.rs:
```diff
@@ -269,7 +269,7 @@ pub struct DAGCircuit {
 
     /// Operation kind to count
-    op_names: IndexMap<String, usize, RandomState>,
+    pub op_names: IndexMap<String, usize, RandomState>,
```
I think instead of making this public we should be using `count_ops(py, false)`

```suggestion
    op_names: IndexMap<String, usize, RandomState>,
```

Although I can see the value here of having a no copy access method to `op_names` that returns a `&IndexMap` instead of an owned copy like `count_ops` does.

mtreinish reviewed (commented):
Sorry I glossed over the changes to `dag_circuit.rs` in my last review. We shouldn't change the interface to `insert_1q_on_incoming_qubit` unless we need to. Since nothing is using it anymore we should revert that piece. Also, I think we shouldn't make `op_names` public because that opens up the possibility of a transpiler pass corrupting the name counts accidentally by mutating it. We have the `count_ops()` method already, but we can also add a no-copy method to get an immutable reference to the inner mapping. These are so minor I'll push up a quick commit to fix both right now. Otherwise I think this is ready to merge.

mtreinish committed:
Fix rustfmt

mtreinish committed:
Don't change interface for DAGCircuit

Previously this PR was making the op_names field of the DAGCircuit
struct public so it was accessible to the new transpiler pass code.
However, this opened up the possibility of mutating the field by
mistake, instead this makes the field private again and adds a no-copy
method to get an immutable reference to the field. Additionally, there
were some interface changes made to one method in DAGCircuit that were
not needed anymore, which this reverts to minimize the diff.

mtreinish committed:
Remove last .first().unwrap()

mtreinish reviewed (approved):
LGTM now, thanks for all the work on this! It's going to have a very positive impact on runtime performance for the transpiler at optimization levels 2 and 3.
